src/bridges/yearn/YearnBridge.sol

```diff
@@ -0,0 +1,205 @@
+
+import {AztecTypes} from "../../aztec/libraries/AztecTypes.sol";
+import {IRollupProcessor} from "../../aztec/interfaces/IRollupProcessor.sol";
+import {IYearnRegistry} from "../../interfaces/yearn/IYearnRegistry.sol";
+import {IYearnVault} from "../../interfaces/yearn/IYearnVault.sol";
+import {IWETH} from "../../interfaces/IWETH.sol";
+import {ErrorLib} from "../base/ErrorLib.sol";
+import {BridgeBase} from "../base/BridgeBase.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+/**
+ * @title Aztec Connect Bridge for depositing on Yearn
+ * @author Majorfi
+ * @notice You can use this contract to deposit tokens on compatible Yearn Vaults
+ */
+contract YearnBridge is BridgeBase {
+    using SafeERC20 for IERC20;
+
+    IYearnRegistry public constant YEARN_REGISTRY = IYearnRegistry(0x50c1a2eA0a861A967D9d0FFE2AE4012c2E053804);
+    address public constant WETH = address(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+
+    error InvalidOutputANotLatest();
+    error InvalidWETHAmount();
+    error InvalidOutputVault();
+    error InvalidVault();
+
+    /**
+     * @notice Set address of rollup processor
+     * @param _rollupProcessor Address of rollup processor
+     */
+    constructor(address _rollupProcessor) BridgeBase(_rollupProcessor) {}
+
+    receive() external payable {}
+
+    /**
+     * @notice Set all the necessary approvals for all the latests vaults for the tokens supported by Yearn
+     */
+    function preApproveAll() external {
+        uint256 numTokens = YEARN_REGISTRY.numTokens();
+        for (uint256 i; i < numTokens; ) {
+            address token = YEARN_REGISTRY.tokens(i);
+            address vault = YEARN_REGISTRY.latestVault(token);
+            _preApprove(vault, token);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /**
+     * @notice Set all the necessary approvals for a given vault and its underlying.
+     * @dev The vault must be a valid Yearn vault, registered in the registry
+     * @param _vault - vault address
+     */
+    function preApprove(address _vault) external {
+        address token = IYearnVault(_vault).token();
+        uint256 numVaultsForToken = YEARN_REGISTRY.numVaults(token);
+        for (uint256 i; i < numVaultsForToken; ) {
+            address vault = YEARN_REGISTRY.vaults(token, i);
+            if (vault == _vault) {
+                _preApprove(vault, token);
+                return;
+            }
+            unchecked {
+                ++i;
+            }
+        }
+        revert InvalidVault();
+    }
+
+    /**
+     * @notice A function which will allow the user to deposit/withdraw a given amount of a given token to a given vault. When depositing to a Yearn vault,
+     * the user deposits the underlying (or want) token as input (ex: DAI). When withdrawing from a Yearn vault, the user asks to withdraws some shares of the
```

Comment:
"the underlying (or want)"
I don't understand the "(or want)" here.

Please also reformulate this:
"the user asks to withdraws some shares"

Reply:
Yearn strategies have a `want` token specifying the input asset, [see](https://github.com/yearn/yearn-vaults/blob/master/contracts/BaseStrategy.sol#L243).

Reply:
Oh ok, my bad. @Majorfi feel free to ignore this then